File: frontend/src/lib/types/monitoring.ts
import type { RiskTier } from "./investigation";

export interface EntityRiskScore {
  entity_id:              string;
  entity_type:            string;
  score_date:             string;
  composite_score:        number;
  risk_tier:              RiskTier;
  finding_velocity:       number;
  exposure_trajectory:    number;
  escalation_probability: number;
  trend_direction:        string;
}

export interface TrendRecord {
  entity_id:          string;
  entity_type:        string;
  rule_code:          string;
  window_type:        string;
  finding_count:      number;
  critical_count:     number;
  risk_score:         number;
  trend_direction:    string;
  velocity:           number;
  acceleration:       number;
  prior_period_count: number;
}

export interface Correlation {
  case_id_a:        string;
  case_id_b:        string;
  correlation_type: string;
  strength:         number;
  explanation:      string;
  shared_entities:  Record<string, unknown>;
}

export interface DriftSignal {
  drift_type:    string;
  subject_id:    string;
  subject_label: string;
  magnitude:     string;
  direction:     string;
  change_pct:    number;
  explanation:   string;
}

export interface MonitoringRun {
  run_id:              string;
  run_type:            string;
  status:              string;
  findings_evaluated:  number;
  new_findings:        number;
  drifts_detected:     number;
  correlations_found:  number;
  started_at:          string | null;
  completed_at:        string | null;
}

export interface IntelligenceSummary {
  as_of:                  string;
  top_risk_entities:      EntityRiskScore[];
  worsening_trends:       TrendRecord[];
  high_correlations:      Correlation[];
  critical_drift_signals: DriftSignal[];
  last_monitoring_run:    MonitoringRun | null;
}

File: api/schemas/monitoring.py
"""Typed API contracts for intelligence monitoring endpoints."""
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class TrendRecordSchema(BaseModel):
    entity_id:          str
    entity_type:        str
    rule_code:          str
    window_type:        str
    finding_count:      int
    critical_count:     int
    risk_score:         float
    trend_direction:    str
    velocity:           float
    acceleration:       float
    prior_period_count: int


class EntityRiskScoreSchema(BaseModel):
    entity_id:              str
    entity_type:            str
    score_date:             date
    composite_score:        float
    risk_tier:              str
    finding_velocity:       float
    exposure_trajectory:    float
    escalation_probability: float
    trend_direction:        str


class CorrelationSchema(BaseModel):
    case_id_a:          str
    case_id_b:          str
    correlation_type:   str
    strength:           float
    explanation:        str
    shared_entities:    dict = Field(default_factory=dict)


class DriftSignalSchema(BaseModel):
    drift_type:    str
    subject_id:    str
    subject_label: str
    magnitude:     str
    direction:     str
    change_pct:    float
    explanation:   str


class MonitoringRunSchema(BaseModel):
    run_id:               UUID
    run_type:             str
    status:               str
    findings_evaluated:   int = 0
    new_findings:         int = 0
    drifts_detected:      int = 0
    correlations_found:   int = 0
    started_at:           datetime | None = None
    completed_at:         datetime | None = None

    model_config = {"from_attributes": True}


class IntelligenceSummaryResponse(BaseModel):
    as_of:                  date
    top_risk_entities:      list[EntityRiskScoreSchema] = Field(default_factory=list)
    worsening_trends:       list[TrendRecordSchema] = Field(default_factory=list)
    high_correlations:      list[CorrelationSchema] = Field(default_factory=list)
    critical_drift_signals: list[DriftSignalSchema] = Field(default_factory=list)
    last_monitoring_run:    MonitoringRunSchema | None = None
